devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
+
+`job_queue` is a module that allows jobs to be executed asynchronously. It queues jobs when there are no requests and queues requests when there are no jobs. When a job arrives and there are queued requests, it executes the least recently queued request with the job. When a request arrives and there are queued jobs, it executes the request with the least recently queued job.
+
+Consumers of `job_queue` must create a "`session`" before they can queue requests. A session can make two kinds of requests:
+
+- `get` requests
+- `pop` requests
+
+The consumer provides a `JOB_REQUEST_FUNCTION` that is executed when that request is executed. If the `JOB_REQUEST_FUNCTION` returns `true`, that job is considered completed, and it's  `JOB_COMPLETE_CALLBACK` is called. Once a job is completed, any new sessions that will be started will not see the completed job.
+
+Even after a job is completed, it is not discarded. A job is discarded when all the sessions that were active when the job was pushed have popped it. If all the sessions that were active when the job was queued have popped a job without indicating that the job is complete, the `JOB_COMPLETE_CALLBACK` is called with `JOB_QUEUE_JOB_RESULT_ERROR` as the result.
+
+Sessions make progress independent of each other. It is possible that one session has popped all the queued requests, whereas another session has not popped any. A session can be thought of as an independent "view" of the job queue. Once a session has called `pop` on a job, it cannot call `pop` or `get` on the same job again. If a session has called `get` on a job, it can call `pop` on that job but cannot call `get` on that `job`.
```

Comment:
>Once a session has called `pop` on a job, it cannot call `pop` or `get` on the same job again. If a session has called `get` on a job, it can call `pop` on that job but cannot call `get` on that `job`. [](http://example.com/codeflow?start=233&length=201)

Is this a restriction on the consumer calling this? Meaning you can't call "get" twice in a row? Or is this an implementation detail that the queue will not give the same job multiple times?

It's worded like the former, but it seems more like the latter is true because there is no such thing below as calling pop **on a job**. It looks like "pop" does not take any job argument so which job is popped is handled internally.

Consider re-wording so it's clear:
- What happens when calling "get" multiple times from the same session in a row?
- What happens when calling "get" then "pop"?
etc.

Reply:
I could have worded this better.

When I say "session calls pop/get on a job" I mean "session submits a pop/get request which gets matched to a job".

Q:What happens when calling "get" multiple times from the same session in a row?

A: The `JOB_REQUEST_FUNCTION` for each get request executes with jobs in the FIFO order of job submission until there are no more jobs left at which point the get requests get queued.

Q: What happens when calling "get" then "pop"?
A: Say two jobs are pushed: job X and job Y in that order. There are 4 active sessions: A, B, C, D.

session A submits get request, receives job X
session A submits get request, receives job Y
session A submits pop request, recieves job X
session A submits pop request, receives job Y

session B submits get request, receives job X
session B submits pop request, recieves job X
session B submits get request, receives job Y
session B submits pop request, receives job Y

session C submits pop request, receives job X
session C submits get request, recieves job Y
session C submits pop request, received job Y
session C submits get request, request is queued

session D submits pop request, receives job X
session D submits pop request, recieves job Y
session D submits get request, job is queued
session D submits get request, request is queued